sdk/tables/azure-data-tables/src/main/java/com/azure/data/tables/implementation/TableBearerTokenChallengeAuthorizationPolicy.java

```diff
@@ -0,0 +1,129 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+package com.azure.data.tables.implementation;
+
+import com.azure.core.credential.TokenCredential;
+import com.azure.core.credential.TokenRequestContext;
+import com.azure.core.http.HttpPipelineCallContext;
+import com.azure.core.http.HttpResponse;
+import com.azure.core.http.policy.BearerTokenAuthenticationPolicy;
+import com.azure.core.util.CoreUtils;
+import reactor.core.publisher.Mono;
+
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Locale;
+import java.util.Map;
+
+/**
+ * A policy that authenticates requests with the Storage Table Service and supports challenges including tenantId
+ * discovery. The content added by this policy is leveraged in {@link TokenCredential} to get and set the correct
+ * "Authorization" header value.
+ *
+ * @see TokenCredential
+ */
+public class TableBearerTokenChallengeAuthorizationPolicy extends BearerTokenAuthenticationPolicy {
+    private static final String BEARER_TOKEN_PREFIX = "Bearer ";
+    private static final String WWW_AUTHENTICATE = "WWW-Authenticate";
+    private String[] scopes;
+    private String tenantId;
+    private boolean enabledTenantDiscovery;
+
+    /**
+     * Creates a {@link TableBearerTokenChallengeAuthorizationPolicy}.
+     *
+     * @param credential The token credential to authenticate the request.
+     */
+    public TableBearerTokenChallengeAuthorizationPolicy(TokenCredential credential, boolean enabledTenantDiscovery,
+                                                        String... scopes) {
+        super(credential, scopes);
+        this.scopes = scopes;
+        this.enabledTenantDiscovery = enabledTenantDiscovery;
+    }
+
+    /**
+     * Extracts attributes off the bearer challenge in the authentication header.
+     *
+     * @param authenticateHeader The authentication header containing the challenge.
+     * @param authChallengePrefix The authentication challenge name.
+     *
+     * @return A challenge attributes map.
+     */
+    private static Map<String, String> extractChallengeAttributes(String authenticateHeader,
```

Comment:
It could be, we'd need to take into account that challenges from different services can be slightly different though. For example, Key Vault separates challenge attributes using a comma and a space (", ") and Storage uses just a space (" "). I wonder if other services have even more ways of doing this.

Any thoughts @alzimmermsft?